controllers/capabilities/proxyconfigpromote_controller.go

```diff
@@ -0,0 +1,230 @@
+	if err != nil {
+		if errors.IsNotFound(err) {
+			reqLogger.Info("resource not found. Ignoring since object must have been deleted")
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, err
+	}
+
+	// get providerAccountRef from product
+	providerAccount, err := controllerhelper.LookupProviderAccount(r.Client(), proxyConfigPromote.GetNamespace(), product.Spec.ProviderAccountRef, r.Logger())
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// connect to the 3scale porta client
+	threescaleAPIClient, err := controllerhelper.PortaClient(providerAccount)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	//if proxyConfigPromote.Status.State != "Completed" {
+	if !proxyConfigPromote.Status.Conditions.IsTrueFor("Ready") {
+		statusReconciler, reconcileErr := r.proxyConfigPromoteReconciler(proxyConfigPromote, reqLogger, threescaleAPIClient, product)
+		statusResult, statusUpdateErr := statusReconciler.Reconcile()
+		if statusUpdateErr != nil {
+			if reconcileErr != nil {
+				return ctrl.Result{}, fmt.Errorf("Failed to reconcile activedoc: %v. Failed to update activedoc status: %w", reconcileErr, statusUpdateErr)
+			}
+
+			return ctrl.Result{}, fmt.Errorf("Failed to update activedoc status: %w", statusUpdateErr)
+		}
+
+		if statusResult.Requeue {
+			return statusResult, nil
+		}
+	}
+	if proxyConfigPromote.Spec.DeleteCR != nil && proxyConfigPromote.Status.Conditions.IsTrueFor("Ready") {
+		err := r.DeleteResource(proxyConfigPromote)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *ProxyConfigPromoteReconciler) proxyConfigPromoteReconciler(proxyConfigPromote *capabilitiesv1beta1.ProxyConfigPromote, reqLogger logr.Logger, threescaleAPIClient *threescaleapi.ThreeScaleClient, product *capabilitiesv1beta1.Product) (*ProxyConfigPromoteStatusReconciler, error) {
+
+	var latestStagingVersion int
+	var latestProductionVersion int
+	//get product
+
+	productList, err := threescaleAPIClient.ListProducts()
+	if err != nil {
+		statusReconciler := NewProxyConfigPromoteStatusReconciler(r.BaseReconciler, proxyConfigPromote, "Failed", "", 0, 0, err)
+		return statusReconciler, err
+	}
+	productID := FindServiceBySystemName(*productList, product.Spec.SystemName)
+
+	if productID > 0 {
+		productIDStr := strconv.Itoa(int(productID))
+
+		if proxyConfigPromote.Spec.Production == nil {
+			// check the existing config to get the lastUpdate time
+			oldConfig, err := threescaleAPIClient.GetLatestProxyConfig(productIDStr, "sandbox")
+			newConfig, err := threescaleAPIClient.DeployProductProxy(*product.Status.ID)
+			if err != nil {
+				statusReconciler := NewProxyConfigPromoteStatusReconciler(r.BaseReconciler, proxyConfigPromote, "Failed", productIDStr, 0, 0, err)
+				return statusReconciler, err
+			}
+			// if the UpdateAt strings are the same return a failed status as no product config changes
+			if oldConfig.ProxyConfig.Content.Proxy.UpdatedAt == newConfig.Element.UpdatedAt {
```

Comment:
If there's nothing to be updated, the deployProductProxy() should return an error however it doesn't. We use the timestamps to make sure that the product has been updated. If there have been no changes made to the product config then the state of proxyConfigPromote should be `Ready: false` as technically we haven't promoted anything.

Reply:
I disagree.  The user creates a ProxyConfigPromote CR because the user wants the latest proxy config to be promoted to staging (and optionally to production). 

If the latest proxy-config is already in staging, the ProxyConfigPromote CR has done the job and should report Ready: true. Optionally you can add a message saying that nothing was changed because it was already a the latest. But for me, it is not an error.